google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,422 @@
+    cluster_id: str | None = None
+    zone: str | None = None
+    completed_attempts: list[CompletedAttemptMetric] = field(default_factory=list)
+    handlers: list[MetricsHandler] = field(default_factory=list)
+    is_streaming: bool = False  # only True for read_rows operations
+    was_completed: bool = False
+
+    @property
+    def state(self) -> OperationState:
+        if self.was_completed:
+            return OperationState.COMPLETED
+        elif self.active_attempt is None:
+            if self.completed_attempts:
+                return OperationState.BETWEEN_ATTEMPTS
+            else:
+                return OperationState.CREATED
+        else:
+            return OperationState.ACTIVE_ATTEMPT
+
+    def start(self) -> None:
+        """
+        Optionally called to mark the start of the operation. If not called,
+        the operation will be started at initialization.
+
+        If the operation was completed or has active attempts, will raise an
+        exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.CREATED:
+            return self._handle_error(INVALID_STATE_ERROR.format("start", self.state))
+        self.start_time = time.monotonic()
+
+    def start_attempt(self) -> None:
+        """
+        Called to initiate a new attempt for the operation.
+
+        If the operation was completed or there is already an active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.BETWEEN_ATTEMPTS and self.state != OperationState.CREATED:
+            return self._handle_error(INVALID_STATE_ERROR.format("start_attempt", self.state))
+
+        self.active_attempt = ActiveAttemptMetric()
+
+    def add_call_metadata(self, metadata: dict[str, bytes | str]) -> None:
+        """
+        Attach trailing metadata to the active attempt.
+
+        If not called, default values for the metadata will be used.
+
+        If the operation was completed or there is no active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+
+        Args:
+          - metadata: the metadata as extracted from the grpc call
+        """
+        if self.state != OperationState.ACTIVE_ATTEMPT:
+            return self._handle_error(INVALID_STATE_ERROR.format("add_call_metadata", self.state))
+
+        if self.cluster_id is None or self.zone is None:
+            bigtable_metadata = metadata.get(BIGTABLE_METADATA_KEY)
+            if bigtable_metadata:
+                try:
+                    decoded = "".join(
+                        c if c.isprintable() else " "
+                        for c in bigtable_metadata.decode("utf-8")
+                    )
+                    split_data = decoded.split()
+                    self.cluster_id = split_data[0]
+                    self.zone = split_data[1]
```

Comment:
consider caching this

Reply:
What do you have in mind? Right now it's saved at the operation level, so the operation will only try to parse these fields if unset. But each operation starts fresh. 

Are there any other guarantees we can take advantage of? Should cluster and zone treated the same?

Reply:
Oh so that's the string "unspecified"? 

I had the "global" default in place, but I had cluster field being literally empty by default